src/Microsoft.DotNet.Wpf/src/PresentationCore/System/Windows/Input/Stylus/Wisp/WispStylusDevice.cs

```diff
@@ -418,52 +418,54 @@ private void UpdateOverProperty(IInputElement oldOver, IInputElement newOver)
-                    ((UIElement3D)o).IsEnabledChanged -= _overIsEnabledChangedEventHandler;
-                    ((UIElement3D)o).IsVisibleChanged -= _overIsVisibleChangedEventHandler;
-                    ((UIElement3D)o).IsHitTestVisibleChanged -= _overIsHitTestVisibleChangedEventHandler;
+                    uie3D.IsEnabledChanged -= _overIsEnabledChangedEventHandler;
+                    uie3D.IsVisibleChanged -= _overIsVisibleChangedEventHandler;
+                    uie3D.IsHitTestVisibleChanged -= _overIsHitTestVisibleChangedEventHandler;
                 }
+                // todo patternmatching: Should we throw here?
             }
             if (_stylusOver != null)
             {
                 o = _stylusOver as DependencyObject;
-                if (InputElement.IsUIElement(o))
+                if (o is UIElement uie)
                 {
-                    ((UIElement)o).IsEnabledChanged += _overIsEnabledChangedEventHandler;
-                    ((UIElement)o).IsVisibleChanged += _overIsVisibleChangedEventHandler;
-                    ((UIElement)o).IsHitTestVisibleChanged += _overIsHitTestVisibleChangedEventHandler;
+                    uie.IsEnabledChanged += _overIsEnabledChangedEventHandler;
+                    uie.IsVisibleChanged += _overIsVisibleChangedEventHandler;
+                    uie.IsHitTestVisibleChanged += _overIsHitTestVisibleChangedEventHandler;
                 }
-                else if (InputElement.IsContentElement(o))
+                else if (o is ContentElement ce)
                 {
-                    ((ContentElement)o).IsEnabledChanged += _overIsEnabledChangedEventHandler;
+                    ce.IsEnabledChanged += _overIsEnabledChangedEventHandler;
 
                     // NOTE: there are no IsVisible or IsHitTestVisible properties for ContentElements.
                     //
-                    // ((ContentElement)o).IsVisibleChanged += _overIsVisibleChangedEventHandler;
-                    // ((ContentElement)o).IsHitTestVisibleChanged += _overIsHitTestVisibleChangedEventHandler;
+                    // ce.IsVisibleChanged += _overIsVisibleChangedEventHandler;
+                    // ce.IsHitTestVisibleChanged += _overIsHitTestVisibleChangedEventHandler;
                 }
-                else
+                else if (o is UIElement3D uie3D)
                 {
-                    ((UIElement3D)o).IsEnabledChanged += _overIsEnabledChangedEventHandler;
-                    ((UIElement3D)o).IsVisibleChanged += _overIsVisibleChangedEventHandler;
-                    ((UIElement3D)o).IsHitTestVisibleChanged += _overIsHitTestVisibleChangedEventHandler;
+                    uie3D.IsEnabledChanged += _overIsEnabledChangedEventHandler;
+                    uie3D.IsVisibleChanged += _overIsVisibleChangedEventHandler;
+                    uie3D.IsHitTestVisibleChanged += _overIsHitTestVisibleChangedEventHandler;
                 }
+                // todo patternmatching: Should we throw here?
```

Comment:
What  to do about this ??

Reply:
Yeah, @batzen, sorry for the open ended comment. It was actually for the team. We have had a discussion regarding this and it would be better to throw an error here. But I am still not sure what error message should be thrown here. Previously, we would have got a Cast Error Exception , should we throw the same error now ??

Reply:
Had a second look and it wouldn't even always be an InvalidCastException but a NullRef one.
The message in the Wisp* and PointerLogic classes would be slightly different to the one in PresentationSource and KeyboardDevice though, as the code in Wisp* and PointerLogic would, previously, also throw if the object wasn't a DependencyObject.

I guess it would be better to add some helper to throw an exception with a dedicated message for those places.
Something that states which types would have been support and which type was received might be the most helpful.

Reply:
My suggestion would be to throw an exception. To me, it looks like an InvalidCastException could happen if oldOver is DependencyObject and is not UIElement, ContentElement or UIElement3D while a NullReferenceException could happen if oldOver is not DependencyObject. Since an exception would already be thrown, throwing a proper exception is probably the right thing to do.

Also, adding an else that should never happen should not impact perf here (The method is probably not inlined already because of its size). You could use copy this throw here:
https://github.com/dotnet/wpf/blob/6e053f99e31de32162f8e6a97f42a04742ed821e/src/Microsoft.DotNet.Wpf/src/PresentationCore/System/Windows/Input/Stylus/Wisp/WispStylusDevice.cs#L280

Reply:
That exception is meant to be thrown when InputElement.IsValid returns false and the input element is not null, so it means something else than the new cases have to express.

Reply:
InputElement.IsValid does the same validation as the pattern matching code here but yes, you could add another throw when the input element is not a DependencyObject.

Reply:
In order to unblock this PR, we can go ahead with the `SRID.Invalid_IInputElement` exception for now and create a new issue to introduce a new exception type ( public API change, would require a complete review process ).

The reason for throwing an error is to ensure that we don't swallow exceptions which would have been the case if we don't throw any exception here.